sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -388,6 +388,21 @@ abstract class BinaryComparison extends BinaryOperator with Predicate {
       defineCodeGen(ctx, ev, (c1, c2) => s"${ctx.genComp(left.dataType, c1, c2)} $symbol 0")
     }
   }
+
+  override def inputType: AbstractDataType = AnyDataType
```

Comment:
We need this to satisfy `BinaryOperator`, the type checking logic is defined by both `BinaryOperator` and `checkInputDataTypes` here. Shall we make `BinaryComparison` not extends `BinaryOperator` and duplicate its type checking logic here?